### PR TITLE
docs: update NG0506

### PR DESCRIPTION
**Pull request by arturovt**
Provide a thorough explanation of why an application may remain unstable and why developers should consider not running asynchronous code on the server outside of the Angular zone, as this may lead to memory leaks.

### CODE REVIEW

File: adev/src/content/reference/errors/NG0506.md [hunk 1]
```diff
@@ -2,19 +2,19 @@
...
+- in a browser this signal is used to start the post-hydration cleanup to remove DOM nodes that remained unclaimed
 
 This warning is displayed when the `ApplicationRef.isStable()` doesn't emit `true` within 10 seconds. If this is intentional and your application becomes stable later, you can ignore this warning.
 
 ## Macrotasks
 
-Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame` etc.
-If one of these functions is called in the initialization phase of the app or the bootstrapped components, the application will remain unstable.
+Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame`, etc.
+If one of these functions is called during the initialization phase of the application or in bootstrapped components, it may lead to the application remains unstable.
```
Comment (JeanMeche) [inline adev/src/content/reference/errors/NG0506.md:17]: Should we make these sentences "ready" for zoneless ?

File: adev/src/content/reference/errors/NG0506.md [hunk 2]
```diff
@@ -2,19 +2,19 @@
 
 This warning occurs when hydration is enabled on the client but the NgZone remains unstable for a long period of time.
 
-The `ApplicationRef.isStable()` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
+The `ApplicationRef.isStable` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
```
Comment (JeanMeche) [inline adev/src/content/reference/errors/NG0506.md:5]: > The `ApplicationRef.isStable` API uses NgZone 
In v18 it uses the internal `PendingTasks` service.

Comment (arturovt) [reply adev/src/content/reference/errors/NG0506.md]: Made updates.

**Comment by pkozlowski-opensource** [comment]
I would expect mention of the `afterRender` / `afterNextRender` hooks in this docs as a place to locate logic that needs to run in a browser (as an alternative to running code outside of zones)

File: adev/src/content/reference/errors/NG0506.md [hunk 3]
```diff
@@ -2,19 +2,23 @@
 
 This warning occurs when hydration is enabled on the client but the NgZone remains unstable for a long period of time.
 
-The `ApplicationRef.isStable()` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
+The `ApplicationRef.isStable` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks. In version 18 and subsequent releases, it uses the internal `PendingTasks` service, with NgZone now contributing to the `PendingTasks`.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:5]: I think we can rename the title to "Application remains unstable", i.e. drop "NgZone", since the problem may technically happen in Zoneless mode too. I'd also propose to drop the "In version 18 and subsequent releases, it uses the internal `PendingTasks` service, with NgZone now contributing to the `PendingTasks`." sentence - it's an internal implementation detail (may not have a lot of value to developers debugging a problem). We can mention though that https://angular.dev/api/core/ExperimentalPendingTasks also contribute to the application stability.

File: adev/src/content/reference/errors/NG0506.md [hunk 4]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
 
-        setTimeout(() => { ... }, 10_000)
-    }
+    setTimeout(() => { ... }, 10_000)
+  }
 }
 ```
 
-If these functions need to be called in the initialization phase, invoking them outside the angular zone solves the issue.  
+Consider evaluating whether these asynchronous tasks should be invoked during server-side rendering, or if they should only be triggered in the browser. Running timers on the server may result in delayed response times, as the Angular framework waits for the application to stabilize.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:40]: This error is thrown in the browser during the hydration process. I'm not sure if mentioning server part of it would be relevant in this guide.

File: adev/src/content/reference/errors/NG0506.md [hunk 5]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
+
+Invoking those functions outside of the Angular zone on the server is considered a "dirty" hack because it may result in "dangling" timers or other tasks, which should be avoided during server-side rendering. The root view is destroyed after the final HTML is serialized, hence any dangling timers may lead to memory leaks.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:42]: Similar to the previous comment - this may not be relevant for this guide (we can find a different location in the docs for this info).

If we end up leaving this paragraph, I'd like to propose an update:
```suggestion
Invoking those functions outside of the Angular zone on the server is not recommended because it may result in "dangling" timers or other tasks, which should be avoided during server-side rendering. The root view is destroyed after the final HTML is serialized, hence any dangling timers may lead to memory leaks.
```

File: adev/src/content/reference/errors/NG0506.md [hunk 6]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
+    });
+  }
+}
+```
+
+In the example above, we used `asapScheduler` when the platform is the server. This approach can significantly reduce the duration of application instability.
+
+### Using `afterNextRender`
+
+You may also resort to using the [`afterNextRender`](api/core/afterNextRender) function, which could provide a cleaner method for executing browser-only code outside of the Angular zone. `afterNextRender` callbacks are automatically invoked outside of the Angular zone and are executed only once, whereas `afterRender` will invoke the provided callback after each render:
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:85]: This behavior has changed recently, the callbacks are no longer invoke outside of NgZone explicitly: https://github.com/angular/angular/commit/7e89753eeff24f52d39fef92600293bf1700cd1b

File: adev/src/content/reference/errors/NG0506.md [hunk 7]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
       inject(NgZone).runOutsideAngular(() => {
         setInterval(() => {}, 1000);
-      })
+      });
     }
+  }
+}
+```
+
+When piping streams with `debounceTime` and specifying a timeout to debounce generated events, it's important to note that this practice may contribute to application instability, particularly during server-side rendering. It's advisable to check the current platform and adjust the timeout settings accordingly on the server side:
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:60]: This also seems to be related to server vs client (see comments above).

File: adev/src/content/reference/errors/NG0506.md [hunk 8]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
+      setInterval(() => {}, 1000);
+    });
+  }
 }
 ```
 
-## Debugging the error
+### Third-party libraries
+
+Third-party libraries can also be a root cause of "fire and forget" asynchronous tasks. Some of these libraries may schedule tasks during their initialization logic, which can prevent Angular from stabilizing properly. For instance, consider LogRocket, a session replay platform, which may set up a timer using `setTimeout` to send network information. The application will remain unstable until the timer fires. It's important to consider wrapping third-party library code with `runOutsideAngular` as well or run the initialization within the `afterNextRender`:
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:99]: I'd propose to avoid referencing a specific library, just mention that some third-party libraries may have patterns that would delay application stability. The recommendation might be the same: try running relevant library code outside of zone.

File: adev/src/content/reference/errors/NG0506.md [hunk 9]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
+    afterNextRender(() => {
+      LogRocket.init(...);
+    });
+  }
+}
+```
+
+### Running code after the hydration is done
+
+You may also delay the code that sets up asynchronous tasks until the hydration process is complete to prevent them from delaying the actual hydration process:
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:125]: That's a good point 👍 I'd propose rephrasing it to say that you can run a code once an application becomes stable (vs "when hydration is done")

File: adev/src/content/reference/errors/NG0506.md [hunk 10]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
+
+```typescript
+import LogRocket from 'logrocket';
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:128]: Similar to one of the previous comments, let's avoid using a specific library.

File: adev/src/content/reference/errors/NG0506.md [hunk 11]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
+        LogRocket.init(...);
+      });
+    }
+  }
+}
+```
+
+## Zoneless applications
+
+Zoneless applications rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:150]:
```suggestion
[Zoneless applications](./guide/experimental/zoneless) rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.
```

File: adev/src/content/reference/errors/NG0506.md [hunk 12]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
+
+In zoneless scenarios, it's more intrinsic to identify the cause of application instability. This might be an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
+
+Developers may also explicitly contribute to indicating the application's stability by using the experimental `PendingTasks` service.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:154]: It'd be great to have a link to API docs for the `ExperimentalPendingTasks` symbol here.

File: adev/src/content/reference/errors/NG0506.md [hunk 13]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
+    }
+  }
+}
+```
+
+## Zoneless applications
+
+Zoneless applications rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.
+
+In zoneless scenarios, it's more intrinsic to identify the cause of application instability. This might be an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:152]:
```suggestion
In zoneless scenarios, stability might be delayed by an application code inside of an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
```

File: adev/src/content/reference/errors/NG0506.md [hunk 14]
```diff
@@ -23,28 +27,130 @@ If one of these functions is called in the initialization phase of the app or th
...
+
+Developers may also explicitly contribute to indicating the application's stability by using the experimental `PendingTasks` service.
 
-Verify that you don't have any long standing microtask or macrotasks.
+In conclusion, it's crucial to monitor all asynchronous tasks scheduled within the application and determine whether they are necessary. Additionally, verifying whether third-party libraries use any asynchronous APIs internally is important, as they may contribute to application instability.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:156]: I'd propose dropping this paragraph, this would mimic the table of contents on a page and info from the first paragraphs.

**Review by AndrewKushnir — COMMENTED** [review]
@arturovt thanks for improving this guide 👍

It looks like some of the recommendations may not be applicable (but some of them might potentially be moved to a different guide), since this error only happens on the client.

**Comment by arturovt** [comment]
@AndrewKushnir I have made updates based on all of your feedback. I removed the parts mentioning server-side rendering. Initially, I included server-side rendering information, but I realized that this warning is specific to browsers only. Additionally, I removed the mention of `afterNextRender` because it is not invoked outside of the zone by default anymore and does not add any value to this guide.

File: adev/src/content/reference/errors/NG0506.md [hunk 15]
```diff
@@ -1,20 +1,24 @@
-# NgZone remains unstable
+# Application remains unstable
 
-This warning occurs when hydration is enabled on the client but the NgZone remains unstable for a long period of time.
+This warning only appears in the browser during the hydration process when it's enabled on the client but the application remains unstable for an extended period.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:3]:
```suggestion
This warning only appears in the browser during the hydration process when it's enabled on the client but the application remains unstable for an extended period of time (over 10 seconds).
```

File: adev/src/content/reference/errors/NG0506.md [hunk 16]
```diff
@@ -1,20 +1,24 @@
...
-Angular Hydration relies on a signal from Zone.js when it becomes stable inside an application:
+Angular Hydration relies on a signal from `ApplicationRef.isStable` when it becomes stable inside an application:
 
-* during the server-side rendering (SSR) to start the serialization process
-* in a browser this signal is used to start the post-hydration cleanup to remove DOM nodes that remained unclaimed
+- during the server-side rendering (SSR) to start the serialization process
+- in a browser this signal is used to start the post-hydration cleanup to remove DOM nodes that remained unclaimed
 
-This warning is displayed when the `ApplicationRef.isStable()` doesn't emit `true` within 10 seconds. If this is intentional and your application becomes stable later, you can ignore this warning.
+This warning is displayed when `ApplicationRef.isStable` does not emit `true` within 10 seconds. If this behavior is intentional and your application stabilizes later, you could choose to ignore this warning. However, it is not recommended to maintain long-running tasks in the background that prevent the application from stabilizing.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:12]: I'd propose removing the last sentence, since it might be intentional (as mentioned earlier in this paragraph):
```suggestion
This warning is displayed when `ApplicationRef.isStable` does not emit `true` within 10 seconds. If this behavior is intentional and your application stabilizes later, you could choose to ignore this warning.
```

File: adev/src/content/reference/errors/NG0506.md [hunk 17]
```diff
@@ -1,20 +1,24 @@
-# NgZone remains unstable
+# Application remains unstable
 
-This warning occurs when hydration is enabled on the client but the NgZone remains unstable for a long period of time.
+This warning only appears in the browser during the hydration process when it's enabled on the client but the application remains unstable for an extended period.
 
-The `ApplicationRef.isStable()` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
+The `ApplicationRef.isStable` API uses NgZone to report when an application becomes `stable` and `unstable`. An application is considered stable when there are no pending microtasks or macrotasks.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:5]: I think we can rephrase this to drop the mention of NgZone (since it's not used in Zoneless mode):
```suggestion
Typically that happens when there are some pending microtasks or macrotasks on a page.
```

File: adev/src/content/reference/errors/NG0506.md [hunk 18]
```diff
@@ -1,20 +1,24 @@
...
 
-This warning is displayed when the `ApplicationRef.isStable()` doesn't emit `true` within 10 seconds. If this is intentional and your application becomes stable later, you can ignore this warning.
+This warning is displayed when `ApplicationRef.isStable` does not emit `true` within 10 seconds. If this behavior is intentional and your application stabilizes later, you could choose to ignore this warning. However, it is not recommended to maintain long-running tasks in the background that prevent the application from stabilizing.
 
-## Macrotasks
+## Applications that use zone.js
 
-Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame` etc.
-If one of these functions is called in the initialization phase of the app or the bootstrapped components, the application will remain unstable.
+Applications that still use zone.js may encounter various factors contributing to instability. These may include pending HTTP requests, timers, or even `fs.readFile` calls, if it's used during server-side rendering on the Node.js platform, as certain Node.js APIs are also partially patched by zone.js.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:16]: I've removed fs and Node-based content, since it's not relevant on the client (where an error happens):
```suggestion
Applications that use zone.js may have various factors contributing to delays in stability. These may include pending HTTP requests, timers (`setInterval`, `setTimeout`) or some logic that continuously invokes `requestAnimationFrame`.
```

File: adev/src/content/reference/errors/NG0506.md [hunk 19]
```diff
@@ -1,20 +1,24 @@
...
+
+### Macrotasks
+
+Macrotasks include functions like `setInterval`, `setTimeout`, `requestAnimationFrame`, etc.
+If one of these functions is called during the initialization phase of the application or in bootstrapped components, it may lead to the application remains unstable.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:21]:
```suggestion
If one of these functions is called during the initialization phase of the application or in bootstrapped components, it may delay the moment when the application becomes stable.
```

File: adev/src/content/reference/errors/NG0506.md [hunk 20]
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
...
 
-        setTimeout(() => { ... }, 10_000)
-    }
+    setTimeout(() => { ... }, 10_000)
+  }
 }
 ```
 
-If these functions need to be called in the initialization phase, invoking them outside the angular zone solves the issue.  
+Consider evaluating whether these asynchronous tasks should be invoked during the bootstrap process. Running timers during bootstrap may result in delayed hydration cleanup times, as the Angular framework waits for the application to stabilize.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:40]: I'd propose removing this sentence, it looks like it repeats the content mentioned earlier.

File: adev/src/content/reference/errors/NG0506.md [hunk 21]
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
...
+
+If you need to schedule those timers in the browser, they should also be wrapped with `runOutsideAngular` to allow for hydration cleanup as soon as possible:
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:42]: We should also rephrase or revert this sentence.

File: adev/src/content/reference/errors/NG0506.md [hunk 22]
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
...
 
 ```typescript
 class SimpleComponent {
-    constructor() {
-      inject(NgZone).runOutsideAngular(() => {
+  constructor() {
+    const platformId = inject(PLATFORM_ID);
+
+    if (isPlatformBrowser(platformId)) {
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:49]: Since the error happens on the client only, there is no need to check for `isPlatformBrowser` here. I'd propose keeping this example simple (similar to the current code).

File: adev/src/content/reference/errors/NG0506.md [hunk 23]
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
...
-      })
+      });
+    }
+  }
+}
+```
+
+### Third-party libraries
+
+Third-party libraries can also be a root cause of "fire and forget" asynchronous tasks. Some third-party libraries may have patterns that could delay application stability. The recommendation remains the same: try executing relevant library code outside of the zone.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:62]:
```suggestion
Some third-party libraries can also produce long-running asynchronous tasks, which may delay application stability. The recommendation is to invoke relevant library code outside of the zone as described above.
```

File: adev/src/content/reference/errors/NG0506.md [hunk 24]
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
...
+### Running code after an application becomes stable
+
+You can run a code that sets up asynchronous tasks once an application becomes stable:
+
+```typescript
+class SimpleComponent {
+  constructor() {
+    const platformId = inject(PLATFORM_ID);
+
+    if (isPlatformBrowser(platformId)) {
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:73]: Do we need `isPlatformBrowser` in this example?

File: adev/src/content/reference/errors/NG0506.md [hunk 25]
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
...
+      });
     }
+  }
 }
 ```
 
-## Debugging the error
+## Zoneless applications
+
+[Zoneless applications](/guide/experimental/zoneless) rely solely on the internal `PendingTasks` service. These pending tasks are explicitly added and removed by the Angular framework. They are used in various scenarios, such as within `effect`s, `@defer` blocks for remote resource loading, during navigation by the router, and by the HTTP module, which marks the application as unstable during HTTP requests.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:89]: This is an implementation detail, I'd propose removing this paragraph.

File: adev/src/content/reference/errors/NG0506.md [hunk 26]
```diff
@@ -23,28 +27,67 @@ If one of these functions is called in the initialization phase of the app or th
...
+
+In zoneless scenarios, stability might be delayed by an application code inside of an `effect` running in an infinite loop (potentially because signals used in effect functions keep changing) or a pending HTTP request.
 
-Verify that you don't have any long standing microtask or macrotasks.
+Developers may also explicitly contribute to indicating the application's stability by using the experimental [`PendingTasks`](/api/core/ExperimentalPendingTasks) service.
```
Comment (AndrewKushnir) [inline adev/src/content/reference/errors/NG0506.md:93]:
```suggestion
Developers may also explicitly contribute to indicating the application's stability by using the experimental [`PendingTasks`](/api/core/ExperimentalPendingTasks) service. If you use the mentioned APIs in your application, make sure you invoke a function to mark the task as completed.
```

**Comment by arturovt** [comment]
@AndrewKushnir made changes.

**Review by AndrewKushnir — APPROVED** [review]
@arturovt thanks for addressing the feedback and improving this page 👍

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 2781b20567f7c717e4d8595997cd61dc88ca7cce.
